File: 0162. Find Peak Element.py
class Solution:
    # recent solution -- iteration
    def findPeakElement(self, nums: List[int]) -> int:
        L, R = 0, len(nums) - 1
        while L < R:
            mid = (L + R) // 2
            if nums[mid] < nums[mid + 1]:
                L = mid + 1
            else:
                R = mid
        return L

    # # previous solution -- recursion
    # def findPeakElement(self, nums: List[int], index=0) -> int:
    #     if len(nums) == 1:
    #         return index
    #     mid = len(nums) // 2
    #     if nums[mid] > nums[mid - 1]:
    #         return index + self.findPeakElement(nums[mid:], mid)
    #     return self.findPeakElement(nums[:mid], index)

File: 0162. Find Peak Element.ts
function findPeakElement(nums: number[]): number {
    let L: number = 0;
    let R: number = nums.length;
    let mid: number = Math.floor((L + R) / 2);

    let left: number | undefined = nums[mid - 1];
    let right: number | undefined = nums[mid + 1];

    // while mid is not peak
    while (
        (left && left > nums[mid])
        || (right && right > nums[mid])
    ) {
        // slopes up to the left
        if (left && left > nums[mid]) {
            R = mid - 1;
        } 
        
        // slopes up to the right
        else if (right && right > nums[mid]) {
            L = mid + 1;
        };

        mid = Math.floor((L + R) / 2);
        left = nums[mid - 1];
        right = nums[mid + 1];
    };

    return mid;
};
